### ISSUE TITLE
Exiting Terminal kills split

### ISSUE THREAD

**Issue by thewatts** [neovim/neovim#5176]
- `nvim --version`:  0.1.4
- Vim (version: ) behaves differently?
- Operating system/version: OSX 10.11.6
- Terminal name/version: iTerm 2 Build 3.0.5
- `$TERM`: xterm-256color

### Actual behaviour
Start: Run Neovim in a vertical split, with a file on each side.
Next: Start a terminal session in the left split.
Next: End that terminal session.
Result: The left split gets closed along with the terminal session.

![terminal-split](https://cloud.githubusercontent.com/assets/1141717/17454855/544de510-5b6b-11e6-9d7b-9bb5ee1de4c2.gif)

### Expected behaviour
Start: Run Neovim in a vertical split, with a file on each side.
Next: Start a terminal session in the left split.
Next: End that terminal session.
Result: The left split stays in-tact, with the previous file back in focus.

This is something that happens normally when closing out of a terminal session in a single window environment, ex:

![terminal-single-window](https://cloud.githubusercontent.com/assets/1141717/17454866/b65b28c6-5b6b-11e6-989f-4282770bd2e4.gif)

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC

Same as above (happens without plugins enabled)
```

**Comment by tweekmonster** [comment]
I think it's because a special buffer is used for the terminal and it's the best way to make sure it's gone after the terminal session is over.

Here's what I wrote to keep the window after the terminal closes:

```vim
function! s:termclose() abort
  let buf = expand('#')
  if !empty(buf) && buflisted(buf) && bufnr(buf) != bufnr('%')
    execute 'autocmd BufWinLeave <buffer> split' buf
  endif
endfunction

autocmd TermClose *:$SHELL,*:\$SHELL call s:termclose()
```

When you run `:terminal` without arguments or with `$SHELL`, it'll switch to the alternate buffer right before it closes.  You may want to look at the full script because I also capture the last two screens of text in a register in case I lose the buffer prematurely:  [term.vim](https://github.com/tweekmonster/dootfiles/blob/master/vim/_source/nvim/99-term.vim)

**Comment by thewatts** [comment]
@tweekmonster thanks for that!

The fact that the terminal buffer is special could definitely be an indicator. However, there doesn't seem to be an issue bringing the previous buffer back into play when only running a single window. I wonder why that would be?

As for the code you provided, the buffer return seems to work like a charm :)! Many Thanks!
Although - on your end, does the original file come back without any syntax highlighting?

**Comment by tweekmonster** [comment]
> Although - on your end, does the original file come back without any syntax highlighting?

Does the highlighting change for you?  I'm not seeing that.  I'm using 0.1.5.

**Comment by thewatts** [comment]
@tweekmonster Yeah, it's odd. If I run `:e` on the file after it's loaded back in - the color comes back.

Currently running 0.1.4

![color-loss](https://cloud.githubusercontent.com/assets/1141717/17458010/d3dd74e8-5bcb-11e6-8774-8fa6044a75e1.gif)

**Comment by tweekmonster** [comment]
Maybe try adding this after the split?

```vim
execute 'doautocmd FileType' buf
```

**Comment by thewatts** [comment]
Hey @tweekmonster! I really appreciate your willingness to help me with this!

When putting the line `executre 'doautocmd FileType' buf` into the function, I get this error:

`E474: Invalid argument: syntax=config/routes.rb` I checked the [`syntax=` function for vim](http://vimdoc.sourceforge.net/htmldoc/syntax.html). It looks like it basically is just taking in a syntax specification.

`config/routes.rb` is the path to the file that is being re-opened in the split.

After the error, the file is still without color (which I assume is because the error itself is blocking it).

My vim-fu isn't very strong - so I apologize if I'm missing something here!

Here's the code I have with your suggested addition:

```
function! s:termclose() abort
  let first = max([1, line('w0') - winheight(0)])
  call setreg('t', getline(first, line('$')), 'V')
  let buf = expand('#')
  if !empty(buf) && buflisted(buf) && bufnr(buf) != bufnr('%')
    execute 'autocmd BufWinLeave <buffer> split' buf
    execute 'doautocmd FileType' buf
  endif
endfunction
```

**Comment by tweekmonster** [comment]
Oh, that was wrong.  Try it with:

```vim
execute 'doautocmd FileType' getbufvar(buf, '&filetype')
```

**Comment by thewatts** [comment]
@tweekmonster really appreciate the help! That wasn't the ticket, unfortunately.

The file still comes up without any color.
I tried to automate the process of calling `:edit` on the file (essentially reloading it - fixing the colors) once the buffer had been loaded - but that didn't work either.

**Comment by thewatts** [comment]
I may have to wait until `0.1.5` ships to see if that's the fix.

**Comment by Shougo** [comment]
0.1.5 is already released.

**Comment by dubiousdavid** [comment]
This is still occurring in 0.1.5 for me.

**Comment by justinmk** [comment]
This is just how `:bdelete` works, it's not unique to terminal buffers. When the `[Process exited ..]` message appears, press `<C-\><C-N>` to get to normal mode, then do whatever you would do to visit another buffer in the same window.

However, I added this to https://github.com/neovim/neovim/issues/5431

**Comment by phmarek** [comment]
I'm wondering about the ratio of times I'd have needed the `:term` output later on (like `:cl` for `:make`), versus how many times it's really just a fire-and-forget.

Would it make sense to keep the buffer around a bit, and/or perhaps recycle it for the same command (a similar one/another one with some identical ID?) instead of allocating fresh buffers all the time?

I've already ended up with buffer numbers ~600, and even if only a few of them are still alive it would be easier to navigate if the buffer numbers were kept as small as possible.

**Comment by tweekmonster** [comment]
@phmarek Check out [neoterm](https://github.com/kassio/neoterm).

**Comment by Cypher1** [comment]
Personally I think this would be great, I don't know why the terms are treated so differently to other buffers.

Thanks guys

**Comment by tssm** [comment]
As @justinmk explained above the terminal buffers are behaving normally. I never realized before that `:bdelete` also closes the buffer window because usually I just `:quit`. Maybe #8975 is The Right Thing to Do? My rationale is that in that way there's no need to handle `:bdelete` in a special way when the buffer is a terminal

**Comment by mlindner** [comment]
Is there any plan to fix this? Really annoying difference from vim's :shell command which returned to the previous buffer. I need to make sure to open a new split first before using :terminal or I lose my position.

**Comment by justinmk** [comment]
@mlindner The plan is to work on this when someone finds time or a patch magically appears. It's currently marked for the 0.4 milestone.

**Comment by gpanders** [comment]
I implemented a patch for this here: https://github.com/neovim/neovim/pull/15442 But there were concerns about backward compatibility with plugins. If somebody else wants to pick up that PR where I left off and address those concerns we should be able to find a solution to this.

**Comment by neovim-discourse** [comment]
This issue has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/stop-terminal-buffer-from-auto-closing-window/3849/1

**Comment by mar-muel** [comment]
@nyngwang 🤔 Simply adding this block doesn't change things for me. Would be nice if you could provide a full code sample?

**Comment by mar-muel** [comment]
Same behavior, unfortunately 😞

**Comment by iam28th** [comment]
If one only needs to mimic Vim's `:shell`, then `:tabe +term` kind of works

**Comment by justinmk** [comment]
`:terminal` now supports `:horizontal` and `:vertical` modifiers, so you can open it in a new split like this:

    :hor te
    :vert te

and when you exit the shell or `:bdelete` it will close that split.

Alternatively, to close a buffer without closing its window, that is tracked in:

- https://github.com/neovim/neovim/issues/2434
